src/iceberg/table_metadata.cc

```diff
@@ -740,6 +898,57 @@ int32_t TableMetadataBuilder::Impl::ReuseOrCreateNewPartitionSpecId(
   return new_spec_id;
 }
 
+int32_t TableMetadataBuilder::Impl::ReuseOrCreateNewSchemaId(
+    const Schema& new_schema) const {
+  // if the schema already exists, use its id; otherwise use the highest id + 1
+  auto new_schema_id = metadata_.current_schema_id.value_or(Schema::kInitialSchemaId);
+  for (auto& schema : metadata_.schemas) {
+    auto schema_id = schema->schema_id().value_or(Schema::kInitialSchemaId);
+    if (schema->SameSchema(new_schema)) {
+      return schema_id;
+    } else if (new_schema_id <= schema_id) {
+      new_schema_id = schema_id + 1;
+    }
+  }
+  return new_schema_id;
+}
+
+Result<std::shared_ptr<PartitionSpec>> TableMetadataBuilder::Impl::UpdateSpecSchema(
+    const Schema& schema, const PartitionSpec& partition_spec) {
+  // UpdateSpecSchema: Update partition spec to use the new schema
+  // This preserves the partition spec structure but rebinds it to the new schema
+
+  // Copy all fields from the partition spec. IDs should not change.
+  std::vector<PartitionField> fields;
+  fields.reserve(partition_spec.fields().size());
+  int32_t last_assigned_field_id = PartitionSpec::kLegacyPartitionDataIdStart;
+  for (const auto& field : partition_spec.fields()) {
+    fields.push_back(field);
+    last_assigned_field_id = std::max(last_assigned_field_id, field.field_id());
+  }
+
+  // Build without validation because the schema may have changed in a way that
+  // makes this spec invalid. The spec should still be preserved so that older
+  // metadata can be interpreted.
+  ICEBERG_ASSIGN_OR_RAISE(auto new_partition_spec,
+                          PartitionSpec::Make(partition_spec.spec_id(), std::move(fields),
+                                              last_assigned_field_id));
+
+  // Validate the new partition name against the new schema
+  ICEBERG_RETURN_UNEXPECTED(new_partition_spec->ValidatePartitionName(schema));
+  return new_partition_spec;
+}
+
+Result<std::unique_ptr<SortOrder>> TableMetadataBuilder::Impl::UpdateSortOrderSchema(
```

Comment:
I think the key difference between C++ and Java impls is that we don't store a `Schema` along with `PartitionSpec` and `SortOrder`. So we even don't need `UpdateSortOrderSchema` and `UpdateSpecSchema` functions, right?